drivers/spi/spi_nrf5.c

```diff
@@ -0,0 +1,339 @@
+ */
+
+#include <spi.h>
+#include <nrfx_spi.h>
+
+#define SYS_LOG_DOMAIN "SPI NRF5"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_SPI_LEVEL
+#include <logging/sys_log.h>
+
+#include "spi_context.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct spi_nrf5_data {
+	struct spi_context ctx;
+	size_t chunk_len;
+	bool busy;
+};
+
+struct spi_nrf5_config {
+	nrfx_spi_t spi;
+};
+
+static inline struct spi_nrf5_data *get_dev_data(struct device *dev)
+{
+	return dev->driver_data;
+}
+
+static inline const struct spi_nrf5_config *get_dev_config(struct device *dev)
+{
+	return dev->config->config_info;
+}
+
+static inline nrf_spi_frequency_t get_nrf_spi_frequency(u32_t frequency)
+{
+	/* Get the highest supported frequency not exceeding the requested one.
+	 */
+	if (frequency < 250000) {
+		return NRF_SPI_FREQ_125K;
+	} else if (frequency < 500000) {
+		return NRF_SPI_FREQ_250K;
+	} else if (frequency < 1000000) {
+		return NRF_SPI_FREQ_500K;
+	} else if (frequency < 2000000) {
+		return NRF_SPI_FREQ_1M;
+	} else if (frequency < 4000000) {
+		return NRF_SPI_FREQ_2M;
+	} else if (frequency < 8000000) {
+		return NRF_SPI_FREQ_4M;
+	} else {
+		return NRF_SPI_FREQ_8M;
+	}
+}
+
+static inline nrf_spi_mode_t get_nrf_spi_mode(u16_t operation)
+{
+	if (SPI_MODE_GET(operation) & SPI_MODE_CPOL) {
+		if (SPI_MODE_GET(operation) & SPI_MODE_CPHA) {
+			return NRF_SPI_MODE_3;
+		} else {
+			return NRF_SPI_MODE_2;
+		}
+	} else {
+		if (SPI_MODE_GET(operation) & SPI_MODE_CPHA) {
+			return NRF_SPI_MODE_1;
+		} else {
+			return NRF_SPI_MODE_0;
+		}
+	}
+}
+
+static inline nrf_spi_bit_order_t get_nrf_spi_bit_order(u16_t operation)
+{
+	if (operation & SPI_TRANSFER_LSB) {
+		return NRF_SPI_BIT_ORDER_LSB_FIRST;
+	} else {
+		return NRF_SPI_BIT_ORDER_MSB_FIRST;
+	}
+}
+
+static int configure(const struct spi_config *spi_cfg)
+{
+	if (spi_cfg->frequency < 125000) {
+		SYS_LOG_ERR("Frequencies lower than 125 kHz are not supported");
+		return -EINVAL;
+	}
+
+	struct spi_context *ctx = &get_dev_data(spi_cfg->dev)->ctx;
+	const nrfx_spi_t *spi = &get_dev_config(spi_cfg->dev)->spi;
+
+	if (!spi_context_configured(ctx, spi_cfg)) {
+		ctx->config = spi_cfg;
+		spi_context_cs_configure(ctx);
+
+		if (spi_cfg->operation & SPI_OP_MODE_SLAVE) {
+			SYS_LOG_ERR("Slave mode is not supported for %s",
+				    spi_cfg->dev->config->name);
+			return -EINVAL;
+		}
+
+		if (spi_cfg->operation & SPI_MODE_LOOP) {
+			SYS_LOG_ERR("Loopback mode is not supported");
+			return -EINVAL;
+		}
+
+		if ((spi_cfg->operation & SPI_LINES_MASK) != SPI_LINES_SINGLE) {
+			SYS_LOG_ERR("Only single line mode is supported");
+			return -EINVAL;
+		}
+
+		if (SPI_WORD_SIZE_GET(spi_cfg->operation) != 8) {
+			SYS_LOG_ERR("Word sizes other than 8 bits"
+				    " are not supported");
+			return -EINVAL;
+		}
+
+		nrf_spi_configure(spi->p_reg,
+				  get_nrf_spi_mode(spi_cfg->operation),
+				  get_nrf_spi_bit_order(spi_cfg->operation));
+	}
+
+	nrf_spi_frequency_set(spi->p_reg,
```

Comment:
this should go into the if {} block above.

Reply:
I thought that frequency could be changed between `transceive` calls. Looks like I misunderstood the "rest of the attributes" in the note describing `spi_config`. I thought it referred to `operation` field only. I'll correct it.